src/main/java/com/yelp/nrtsearch/server/luceneserver/search/SearchRequestProcessor.java

```diff
@@ -0,0 +1,284 @@
+import com.yelp.nrtsearch.server.luceneserver.search.collectors.ScoreCollector;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.lucene.facet.DrillDownQuery;
+import org.apache.lucene.facet.taxonomy.SearcherTaxonomyManager;
+import org.apache.lucene.queryparser.classic.MultiFieldQueryParser;
+import org.apache.lucene.queryparser.classic.QueryParserBase;
+import org.apache.lucene.queryparser.simple.SimpleQueryParser;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.QueryBuilder;
+
+/**
+ * Class to process a {@link SearchRequest} grpc message into the data structures required for
+ * search. Produces a {@link SearchContext} usable to perform search operations.
+ */
+public class SearchRequestProcessor {
+  /**
+   * By default we count hits accurately up to 1000. This makes sure that we don't spend most time
+   * on computing hit counts
+   */
+  public static final int TOTAL_HITS_THRESHOLD = 1000;
+
+  private static final QueryNodeMapper QUERY_NODE_MAPPER = new QueryNodeMapper();
+
+  private SearchRequestProcessor() {}
+
+  /**
+   * Create a {@link SearchContext} representing the given {@link SearchRequest} and index/searcher
+   * state.
+   *
+   * @param searchRequest grpc request message
+   * @param indexState index state
+   * @param shardState shard state
+   * @param searcherAndTaxonomy index searcher
+   * @param diagnostics container message for returned diagnostic info
+   * @return context info needed to execute the search query
+   * @throws IOException if query rewrite fails
+   */
+  public static SearchContext buildContextForRequest(
+      SearchRequest searchRequest,
+      IndexState indexState,
+      ShardState shardState,
+      SearcherTaxonomyManager.SearcherAndTaxonomy searcherAndTaxonomy,
+      SearchResponse.Diagnostics.Builder diagnostics)
+      throws IOException {
+
+    MutableSearchContext context =
+        new MutableSearchContext(
+            indexState, shardState, searcherAndTaxonomy, SearchResponse.newBuilder());
+
+    context.setTimestampSec(System.currentTimeMillis() / 1000);
+    context.setStartHit(searchRequest.getStartHit());
+
+    context.setQueryFields(getQueryFields(context, searchRequest));
+
+    Query query = extractQuery(context.indexState(), searchRequest);
+    diagnostics.setParsedQuery(query.toString());
+
+    Query rewrittenQuery = context.searcherAndTaxonomy().searcher.rewrite(query);
+    diagnostics.setRewrittenQuery(rewrittenQuery.toString());
+
+    Query drillDownQuery = addDrillDowns(context.indexState(), rewrittenQuery);
+    diagnostics.setDrillDownQuery(drillDownQuery.toString());
+
+    context.setQuery(drillDownQuery);
+    context.setCollector(buildCollector(context, searchRequest));
+    context.searchResponse().setDiagnostics(diagnostics);
```

Comment:
it is good that we have access to a mutable searchResponse within the searchcontext. Maybe we need to change DrillSidewaysImpl to accept searchReponse in its constructor to fill out the  facetResult list.